Src/Newtonsoft.Json/Utilities/CollectionUtils.cs

```diff
@@ -340,9 +344,27 @@ public static Array ToMultidimensionalArray(IList values, Type type, int rank)
             }
 
             Array multidimensionalArray = Array.CreateInstance(type, dimensions.ToArray());
-            CopyFromJaggedToMultidimensionalArray(values, multidimensionalArray, new int[0]);
+            CopyFromJaggedToMultidimensionalArray(values, multidimensionalArray, ArrayEmpty<int>());
 
             return multidimensionalArray;
         }
+
+        // 4.6 has Array.Empty<T> to return a cached empty array. Lacking that in other
+        // frameworks, Enumerable.Empty<T> happens to be implemented as a cached empty
+        // array in all versions (in .NET Core the same instance as Array.Empty<T>).
+        // This includes the internal Linq bridge for 2.0.
+        // Since this method is simple and only 11 bytes long in a release build it's
+        // pretty much guaranteed to be inlined, giving us fast access of that cached
+        // array. With 4.5 and up we use AggressiveInlining just to be sure, so it's
+        // effectively the same as calling Array.Empty<T> even when not available.
```

Comment:
I flip-flopped on that a few times when doing this. Having contributed to System.Linq in .Net Core I'd say it would be extremely unlikely in Core and even less likely in a successor to 4.6. (We do even have an enumerable type that in some ways would be better to return in Core, but it's not going to happen).

But there's always the possibility of a completely new version of .NET apart from Core, Mono, the MS closed-source frameworks and it could do something different. Trying a more defensive version I found:

1. Good: It still gets inlined even without the `MethodImplOptions.AggressiveInlining`. (It ends up being 21 bytes of IL with one branch, so still very small and simple).
2. Bad: The null-check isn't optimised away by the jitter.

An extra null-check is probably worth it for the safety though, but I'm leaving the assertion in the debug builds as if such a case did happen this would still be sub-optimal.